tools/jit/gen_unboxing.py

```diff
@@ -0,0 +1,169 @@
+# Generates CodegenUnboxingWrappers.cpp.
+# This generates static unboxing wrapper for ATen ops.
+import argparse
+import json
+import os
+from dataclasses import dataclass
+from tools.codegen.api import unboxing
+from tools.codegen.api.types import CppSignatureGroup
+from tools.codegen.context import method_with_native_function
+from tools.codegen.gen import parse_native_yaml
+from tools.codegen.model import NativeFunction, NativeFunctionsGroup
+from tools.codegen.utils import Target, FileManager, mapMaybe, make_file_manager
+from typing import Union, Sequence
+from typing_extensions import Literal
+
+
+# Generates CodegenFunctions.h & CodegenFunctions.cpp.
+@dataclass(frozen=True)
+class ComputeUnboxingFunctions:
+    target: Union[Literal[Target.DECLARATION], Literal[Target.DEFINITION]]
+
+    @method_with_native_function
+    def __call__(self, f: NativeFunction) -> str:
+        sig_group = CppSignatureGroup.from_native_function(
+            f, method=False, fallback_binding=f.manual_cpp_binding
+        )
+        sig = sig_group.most_faithful_signature()
+
+        if self.target is Target.DECLARATION:
+            # Note [The ATen Codegen Unboxing API]
+            # Similar to the ATen Operators API, ATen Codegen Unboxing API lives in the at::unboxing namespace, and
+            # will be used by codegen unboxing wrappers (CodegenUnboxingWrappers.cpp).
+            # The Wrappers will be registered into torch::jit::OperatorRegistry using RegisterOperators API.
+            #
+            # Important characteristics about the Codegen Unboxing API:
+            # (1) It follows the OperatorRegistry API.
+            #     This is kind of necessary to avoid overhead.
+            #     For example: if it followed the C++ API, then all of the faithful C++ factory functions
+            #     would need to wrap their arguments into TensorOptions only to unwrap them again.
+            # (2) Under the hood it calls C++ API.
+            return f"""
+// aten::{f.func}
+TORCH_API void {f.func.name.unambiguous_name()}(Stack & stack);
+"""
+        else:
+            # gather all the arguments from native function, including "out"
+            args = list(f.func.arguments.flat_non_out) + (
+                list(f.func.arguments.out) if f.func.arguments.out else []
+            )
+
+            # parse arguments into C++ code
+            arguments = unboxing.convert_arguments(
+                args, f.func.arguments.tensor_options
```

Comment:
Why do you need to deal with / worry about TensorOptions?

Since IValues don't store TensorOptions, and the C++ frontend API has versions of each op that don't take in TensorOptions, can you get away with just converting each argument from an IValue directly without having to:

(a) pay the runtime cost of wrapping everything in a `TensorOptions` object
(b) deal with added complexity in the codegen :)

Reply:
Good point, I merely removed the docs about `TensorOptions` but didn't remove the code. I'll do it.